### PR TITLE
[`flake8-slots`] Add plugin, add `SLOT000`, `SLOT001` and `SLOT002`

### PR DESCRIPTION
**Pull request by qdegraaf**
## Summary

Adds all rules from the [flake8-slots](https://github.com/python-formate/flake8-slots) plugin.

No autofixes yet but maybe a `Fix::suggested` autofix with a default `__slots__` val could be useful? 

## Test Plan

Added fixtures for all rules. 

I added a check for "indirect" subclassing in situations like:
```python
from collections import namedtuple 

some_named_tuple = namedtuple("foo", ["bar", "bla"])

class Foo(some_named_tuple):
    pass
```

As I imagined this would happen regularly. For completeness' sake it is arguable we should do this for `str` and `tuple` too in scenarios like:

```python
str_alias = str

class Foo(str_alias):
    pass
```

This will be more rare, left it out for now. Unsure if we have a convention for `base` checks across Ruff (e.g. do we **always** check for type aliases or **never** check for type aliases because XYZ). Happy to add the check to SLOT000 and SLOT001 and/or remove the check from SLOT002 depending on what is preferred. 

NOTE: If adding the check for the other subclasses makes sense. Maybe a general helper which combines some of the logic in `no_slots_in_namedtuple_subclass` and `is_indirect_namedtuple_subclass` and making it more generic so that it becomes something like: `fn is_indirect_subclass_of(expr: &ExprName, subclass: [&str]) -> bool {}` might be useful. 

## Issue links

Closes: https://github.com/charliermarsh/ruff/issues/3873

### CODE REVIEW

File: crates/ruff/src/rules/flake8_slots/rules/no_slots_in_str_subclass.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    F: FnOnce() -> TextRange,
+{
+    if class.bases.len() != 1 {
+        return;
+    }
+
+    let Expr::Name(ast::ExprName { id, .. }) = &class.bases[0] else {
+        return;
+    };
+
```
Comment (MichaReiser) [inline crates/ruff/src/rules/flake8_slots/rules/no_slots_in_str_subclass.rs:56]: NIt
```suggestion
   let [Expr::Name(ast::ExprName { id, .. })] = &class.bases.as_slice() else {
        return;
    };

```

File: crates/ruff/src/rules/flake8_slots/rules/no_slots_in_namedtuple_subclass.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    locate: F,
+) where
+    F: FnOnce() -> TextRange + Copy,
+{
+    if class.bases.len() != 1 {
+        return;
+    }
+
+    if let Expr::Name(ast::ExprName { id, .. }) = &class.bases[0] {
+        let scope = checker.semantic_model().scope();
```
[the remark this answers is not in the document]
Comment (qdegraaf) [reply crates/ruff/src/rules/flake8_slots/rules/no_slots_in_namedtuple_subclass.rs:64]: Ah that's nice! Really loving the extent to which those `if let`'s can check things in an implicit-but-transparent manner.

File: crates/ruff/src/checkers/ast/mod.rs [hunk 3]
```diff
@@ -815,6 +817,24 @@ where
                 if self.enabled(Rule::DuplicateBases) {
                     pylint::rules::duplicate_bases(self, name, bases);
                 }
+
+                if self.enabled(Rule::NoSlotsInStrSubclass) {
+                    flake8_slots::rules::no_slots_in_str_subclass(self, class_def, || {
+                        helpers::identifier_range(stmt, self.locator)
+                    });
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff/src/checkers/ast/mod.rs:824]: We have `AnyNodeRef` and `AnyNode` that can represent arbitrary nodes and you can convert from `Stmt` or `StmtClass` to either of those. We should probably start using those more widely but I don't consider this part of this PR's scope. CC: @charliermarsh

[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/ruff/src/checkers/ast/mod.rs:827]: I think I prefer just passing in the `Stmt`, it's more consistent with how we solve this elsewhere (even if it's not ideal).

File: crates/ruff/src/rules/flake8_slots/rules/no_slots_in_tuple_subclass.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+/// SLOT001
+pub(crate) fn no_slots_in_tuple_subclass<F>(checker: &mut Checker, class: &StmtClassDef, locate: F)
+where
+    F: FnOnce() -> TextRange,
+{
+    if class.bases.len() != 1 {
+        return;
+    }
```
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_slots/rules/no_slots_in_tuple_subclass.rs:51]: I think the originating plugin checks all bases.

File: crates/ruff/src/rules/flake8_slots/rules/no_slots_in_namedtuple_subclass.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                                        }
+                                    }
+                                }
+                            }
+                            _ => (),
+                        }
+                    }
+                }
+            }
+        }
```
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_slots/rules/no_slots_in_namedtuple_subclass.rs:95]: Does the originating plugin do this? As far as I can tell, it doesn't.

Comment (charliermarsh) [reply crates/ruff/src/rules/flake8_slots/rules/no_slots_in_namedtuple_subclass.rs:95]: Gonna simplify for now to bias towards merging, but happy to revisit and we'll have it here in the Git history if we want to restore.

**Pull request merged**